src/PandoraMonitoring.cc

```diff
@@ -991,6 +996,49 @@ void PandoraMonitoring::ViewEvent()
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+void PandoraMonitoring::SaveAndViewEvent(const std::string savePath)
+{
+    // TODO: We should be checking the path exists/check SavePic is capable of
+    // dealing with the case it isn't.
```

Comment:
I've got this TODO sorted now, but annoyingly the C++ `mkdir` doesn't have the equivalent of a  `-p` flag (I.e make folder with all missing parents), which means a larger function is needed.

Do you have a preference where that sort of code should live? I've stuck it next to the save for now, but wasn't sure where a more appropriate location would be / a new file.

Reply:
Hi Ryan, Is this mkdir functionality portable, or actually starting to become OS-specific? We should ensure it's OK on MacOS as well as our usual favourite linux distributions, with gcc and clang. If there's a risk it's not portable, it might be cleaner to see if there's a portable way to just check whether a directory exists, then print a message and raise an exception if it isn't (or event to just try to save and catch/rethrow any exception). What do you think?

Reply:
To my knowledge it works on Linux/macOS, but not Windows.

I've tested it locally on my mac, with Catalina and a Clang/Cmake/Ninja build, as well as on a Uni CentOS7 machine using GCC/Cmake/Ninja, and it worked fine for each. I'll do some more digging tomorrow to check it works on some other common flavours of Linux, and go from there.

If it looks to be a bit too platform dependant, I'll swap it out to alerting the user. I was hoping it was something ROOT might have implemented, but the "solution" they seem to use is using `gSystem` to call out to the system prompt and call `mkdir -p`, which seems a little janky.

Reply:
Thanks very much for testing so thoroughly on different platforms! This is PandoraMonitoring, so we actually have a good deal of flexibility, provided that the others who build this library (usually as one of their external packages) have no issue maintaining their builds.

I'd suggest that, if the error handling is very clean and the reporting is nice and clear without the mkdir functionality, then it's fine to just leave this functionality out.

If the error handling right now (when you try to save a file to a directory that doesn't exist) makes a mess, or the error messages are impenetrable, that we add this functionality, provided that it's definitely portable.

What do you think?

I can take a look at the diffs again if you decide the mkdir functionality should go in. From a quick glance now, I suggest modify function names to be in CamelCase and (starting with a capital letter) and to add dividing comment line between function implementations.

Thanks, and hopefully we can get this in soon!

Reply:
>I'd suggest that, if the error handling is very clean and the reporting is nice and clear without the mkdir functionality, then it's fine to just leave this functionality out.
>
>If the error handling right now (when you try to save a file to a directory that doesn't exist) makes a mess, or the error messages are impenetrable, that we add this functionality, provided that it's definitely portable.

Without the `mkdir` functionality, we just fail silently I think.

I've checked the man page for both `mkdir` and `stat` (https://linux.die.net/man/2/mkdir and https://linux.die.net/man/2/stat) and they both confirm to the POSIX standard, so it should be fine for any POSIX compliant OS.

Reply:
Thanks for checking this.